cpp/src/arrow/record_batch.cc

```diff
@@ -393,7 +393,7 @@ Result<std::shared_ptr<RecordBatchReader>> RecordBatchReader::Make(
 RecordBatchReader::~RecordBatchReader() {
   auto st = this->Close();
   if (!st.ok()) {
-    ARROW_LOG(WARNING) << "Implicitly called RecordBatchReader::Close failed: " << st;
+    st.Warn("Implicitly called RecordBatchReader::Close failed");
```

Comment:
I meant to ask earlier: if `ARROW_WARN_NOT_OK` only takes an expr, then to print a custom message you need to call `Warn()` directly; should I add another macro, something like `ARROW_WARN_MSG_NOT_OK` to also take a message, or should we drop that custom message?

I assume you can't overload a macro, not sure if there's a preference on an approach.

Reply:
Ah, right. Since the current macro isn't used in many places I would be in favor of just making the macro always require a message.

Reply:
cc @vibhatha

Reply:
@drin it would be good to add a test case like that. I am definitely sure, there could have been a good use case when this JIRA was created. We can gather some thoughts from there too.

cc @pitrou

Reply:
So, I looked at why ARROW-16515 was created and it was related to `SubstraitExecutor` and keeping a reference to it. The destructor in that class uses `ARROW_CHECK_OK` which is similar to `ARROW_WARN_NOT_OK` but logs at the `FATAL` level instead. I don't see a clear test case for `Warn()` in that context.

For now, I created a test class that extends RecordBatchReader and returns non-OK in it's Close() method as a simple, lightweight approach to exercising this. I am not sure how important it is to validate the logging mechanism itself, but at least this adds code coverage of the macro and a "failing" Close() method called from a constructor.

Reply:
Well, there's another PR that's just removing the log instead…see https://github.com/apache/arrow/pull/13375/files#diff-0078df867c0f64dcfe79c149ee007f0933d164a32f122e28d79872a969b5be05